File: algorithms/377. Combination Sum IV/solution.py
class Solution(object):
    def combinationSum4(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        # refer to http://www.cnblogs.com/grandyang/p/5705750.html
        dp = [0] * (target+1)
        dp[0] = 1
        for i in range(target+1):
            for n in nums:
                if n + i <= target:
                    dp[n+i] += dp[i]
        return dp[target]

File: algorithms/377. Combination Sum IV/solution.cc
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        vector<int> dp(target + 1);
        dp[0] = 1;
        sort(nums.begin(), nums.end());
        for ( int i = 1; i <= target; ++i ) {
            for ( int j = 0; j < nums.size() && nums[j] <= i; ++j ) {
                dp[i] += dp[i - nums[j]];
            }
        }
        return dp[target];
    }
};
